cpp/src/spatial/hausdorff.cu

```diff
@@ -183,19 +103,35 @@ struct hausdorff_functor {
     auto result_temp_iter = static_cast<hausdorff_acc<T>*>(result_temp.data());
 
     auto scatter_map = thrust::make_transform_iterator(
-      hausdorff_index_iter, [] __device__(hausdorff_index acc) { return acc.result_idx; });
+      gcp_iter, [num_spaces] __device__(cartesian_product_group_index idx) {
+        // the given output is only a "result" if it is the last output for a given pair-of-spaces
```

Comment:
Wouldn't this be true for any use of the cartesian_product_group_index? It's not specific to Hausdorff, right? If not, can it be factored out, too?

Reply:
It's useful for Hausdorff only because we do not have an `accumulate_by_key` and instead are relying on a bastardization of `inclusive_scan_by_key` and some weird scatter output iterator. The other distance metrics can use an _actual_ reduce function (`reduce_by_key`) and therefore don't need this hack (since those reduce operators are commutative).